### ISSUE TITLE
Expanding coderef need to at least be documented as Incompatible Change

### ISSUE THREAD

**Issue by perlancar** [dagolden/Log-Any#36]
In 0.33-TRIAL, there is this Changes entry:

    * Added

    The default formatter now expands code references. If the first argument is a code reference, it is expanded and returned. If an argument to "sprintf" style formatting is a code reference, it is expanded.

First of all, "expand" is not a very exact term. 

Second, previously one assumes that it's safe to dump objects/references including coderefs to be viewed only. But now the logger will execute them. This is certainly unexpected and a backward incompatible change.

Third, I hope you reconsider. I personally do not prefer this String::Flogger-like behavior.

**Comment by xdg** [comment]
I'm happy to clarify "expand" and to raise awareness of the change.
"Lazy" logging with subroutines that don't get executed when there is no logging has been a frequently requested feature, so I want to keep that.

I suspect that people putting coderefs in their logs is uncommon, as they don't provide useful information.  E.g.

```
$ perl -MLog::Any::Adapter::Util=dump_one_line -wE 'say dump_one_line( sub { "hello" } )'
sub { "DUMMY" }
```

But I can vaguely imagine a scenario where someone is dumping subroutine arguments, one of which could be a code reference.

How about this: the subroutine will only be executed if it is the first argument to the 'xxxf' functions (replacing the format string).  That should be backwards compatible.  Anyone who wants something more advanced can write their own formatter.

**Comment by perlancar** [comment]
> I suspect that people putting coderefs in their logs is uncommon, as they don't provide useful information.

I suspect the same. Although sometimes I do deparse the coderefs to see stuffs (e.g. using Log::Any::Proxy::Patch::UseDataDmp).

 > How about this: the subroutine will only be executed if it is the first argument to the 'xxxf' functions (replacing the format string). That should be backwards compatible. Anyone who wants something more advanced can write their own formatter.

If I have to choose between the 'xxx' and 'xxxf' functions for this, I'd probably vote for the 'xxx' function (though I personally don't want to add extra behavior to the established API).

Another alternative is a separate set of say 'xxxc' functions (warnc, debugc, ...) that specifically does this coderef-calling feature. 

I'm still not convinced though that this "lazy logging" warrants an extra semantic. What's wrong with the good old:

    if ($log->is_debug) {
        # do something expensive
    }

Yes, it's more verbose, but does one need lazy logging that often?

**Comment by xdg** [comment]
> If I have to choose between the 'xxx' and 'xxxf' functions for this, I'd probably vote for the 'xxx' function > (though I personally don't want to add extra behavior to the established API).

The reason it goes on the 'xxxf' functions is that this is a "formatting" behavior, and that's where formatter is called.

> I'm still not convinced though that this "lazy logging" warrants an extra semantic.

It's a "feature" of other logging systems.  Some people prefer the brevity.

**Comment by xdg** [comment]
Fixed with 9107dabfd86b4ccdf33bc1fa132c5f81f28a7038
